File: Observer-Pattern/Task -1/180042111.java
import java.util.*;
import java.text.*;

interface ClassTime{
    public void notifyObserver();
    public void AddDisplay(Display display);
    public void RemoveDisplay(Display display);
}

interface Display{
    public void update(String CourseNo, String TopicName, String TeacherName, String date);
    public void displayData();
}

class ClassTimeData implements ClassTime{
    public ArrayList<Display>displays;
    
    public String CourseNo ;
    public String TopicName;
    public String TeacherName;
    public String date;

    ClassTimeData(){
        displays = new ArrayList<Display>();
    }

    @Override
    public void AddDisplay(Display display) {
        displays.add(display);
    }

    @Override
    public void RemoveDisplay(Display display) {
        displays.remove(display);
    }

    public void SetChanges (String CourseNo, String TopicName, String TeacherName, String date){
        this.CourseNo = CourseNo;
        this.TopicName = TopicName;
        this.TeacherName = TeacherName;
        this.date = date;
        notifyObserver();
    }
    
    @Override
    public void notifyObserver() {
        //ServiceLoader<Display> loader = ServiceLoader.load(Display.class);
        for (Display i : displays) {
            i.update(CourseNo, TopicName, TeacherName, date);
        }
    }
}

class TeacherDisplay implements Display {
    public String CourseNo ;
    public String TopicName;
    public String TeacherName;
    public String Classdate;

    SimpleDateFormat dateFormat  = new SimpleDateFormat("dd.MM.yyyy, HH:mm");
    SimpleDateFormat sdf = new SimpleDateFormat("E MMM dd HH:mm:ss z yyyy");

    public ClassTimeData classTimeData;

    public TeacherDisplay (ClassTimeData classTimeData){
        this.classTimeData = classTimeData;
        classTimeData.AddDisplay(this);
    }

    @Override
    public void displayData() {
        System.out.println("Teacher's Display:");
        System.out.println("Course No: " + CourseNo);
        System.out.println("Topic Name: " + TopicName);
        System.out.println("Teacher Name: " + TeacherName);
        System.out.println("Date: " + Classdate);
    }
    @Override
    public void update(String CourseNo, String TopicName, String TeacherName, String date) {
        this.CourseNo = CourseNo;
        this.TopicName = TopicName;
        this.TeacherName = TeacherName;
        
        try {
            Date currentdate = sdf.parse(date);
            this.Classdate = dateFormat.format(currentdate);
        }
        catch (Exception e) {
            e.printStackTrace();
        }
        
        this.displayData();
    }
}

class StudentDisplay implements Display {
    public String CourseNo ;
    public String TopicName;
    public String TeacherName;
    public String Classdate;

    SimpleDateFormat dateFormat = new SimpleDateFormat("dd/MM/yyyy, hh:mm a");
    SimpleDateFormat sdf = new SimpleDateFormat("E MMM dd HH:mm:ss z yyyy");

    public ClassTimeData classTimeData;

    public StudentDisplay (ClassTimeData classTimeData){
        this.classTimeData = classTimeData;
        classTimeData.AddDisplay(this);
    }

    @Override
    public void displayData() {
        System.out.println("Student's Display:");
        System.out.println("Course No: " + CourseNo);
        System.out.println("Topic Name: " + TopicName);
        System.out.println("Teacher Name: " + TeacherName);
        System.out.println("Date: " + Classdate);
        
    }
    @Override
    public void update(String CourseNo, String TopicName, String TeacherName, String date) {
        this.CourseNo = CourseNo;
        this.TopicName = TopicName;
        this.TeacherName = TeacherName;
        
        try {
            Date currentdate = sdf.parse(date);
            this.Classdate = dateFormat.format(currentdate);
        }
        catch (Exception e) {
            e.printStackTrace();
        }

        this.displayData();
    }
}

class NoticeBoard{
    public static void main(String[] args) {

        SimpleDateFormat sdf = new SimpleDateFormat("E MMM dd HH:mm:ss z yyyy");
        String date = sdf.format(new Date());

        ClassTimeData classTimeData = new ClassTimeData();

        new TeacherDisplay(classTimeData);
        new StudentDisplay(classTimeData);
        classTimeData.SetChanges("Design Pattern", "Observer Design Pattern", "Nazmul Haque", date);
    }
} 

File: Observer-Pattern/Task -1/180042111.js
var ClassTimeData = /** @class */ (function () {
    function ClassTimeData() {
        this.displays = [];
        this.getCourseNo = function () { return 'Swe 45xx'; };
        this.getTopicName = function () { return 'xyz'; };
        this.getTeacherName = function () { return 'ABC'; };
        this.getDate = function () { return new Date(); };
    }
    ClassTimeData.prototype.addObserver = function (display) {
        this.displays.push(display);
    };
    ClassTimeData.prototype.removeObserver = function (display) {
        this.displays.filter(function (obj) { return obj !== display; });
    };
    ClassTimeData.prototype.notifyObserver = function () {
        var _this = this;
        this.displays.map(function (o) {
            o.update(_this.CourseNo, _this.TopicName, _this.TeacherName, _this.date);
        });
    };
    ClassTimeData.prototype.measureChange = function (courseNo, time, topicName, teacherName) {
        this.CourseNo = courseNo;
        this.date = time;
        this.TopicName = topicName;
        this.TeacherName = teacherName;
        this.notifyObserver();
    };
    return ClassTimeData;
}());
var TeacherDisplay = /** @class */ (function () {
    function TeacherDisplay() {
    }
    TeacherDisplay.prototype.update = function (CourseNo, TopicName, TeacherName, date) {
        this.CourseNo = CourseNo;
        this.TopicName = TopicName;
        this.TeacherName = TeacherName;
        this.date = date;
        this.display();
    };
    TeacherDisplay.prototype.display = function () {
        console.log("Teacher's Notice");
        console.log("Course No.: ", this.CourseNo);
        console.log("Topic Name: ", this.TopicName);
        console.log("Teacher Name: ", this.TeacherName);
        console.log("Class Time: ", this.date);
    };
    return TeacherDisplay;
}());
var StudentDisplay = /** @class */ (function () {
    function StudentDisplay() {
    }
    StudentDisplay.prototype.update = function (CourseNo, TopicName, TeacherName, date) {
        this.CourseNo = CourseNo;
        this.TopicName = TopicName;
        this.TeacherName = TeacherName;
        this.date = date;
        this.display();
    };
    StudentDisplay.prototype.display = function () {
        console.log("Students Notice:");
        console.log("Course No.: ", this.CourseNo);
        console.log("Topic Name: ", this.TopicName);
        console.log("Teacher Name: ", this.TeacherName);
        console.log("Class Time: ", this.date);
    };
    return StudentDisplay;
}());
var classTimeData = new ClassTimeData();
classTimeData.addObserver(new StudentDisplay());
classTimeData.addObserver(new TeacherDisplay());
classTimeData.measureChange('SWE-4501', new Date(), 'Observer Pattern', 'Nazmul Haque');
